File: QuickUIMate/src/main/java/com/tomcan/frame/vm/QuickViewModel.java
package com.tomcan.frame.vm;

import android.app.Application;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;

import com.tomcan.frame.m.QuickModel;
import com.tomcan.frame.obs.IBaseLifecycle;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * @author TomCan
 * @description:
 * @date :2020/8/5 15:34
 */
public abstract class QuickViewModel<M extends QuickModel> extends AndroidViewModel implements IBaseLifecycle {
    public String TAG = this.getClass().getSimpleName();
    public final Application application;
    public M model;

    public QuickViewModel(@NonNull Application application) {
        super(application);
        this.application = application;
    }

    @Override
    public void onAny(LifecycleOwner owner, Lifecycle.Event event) {
        Log.i(TAG, "onAny");
    }

    @Override
    public void onCreate() {
        ParameterizedType genericSuperclass = (ParameterizedType) getClass().getGenericSuperclass();
        assert genericSuperclass != null;
        Type[] actualTypeArguments = genericSuperclass.getActualTypeArguments();
        Class<M> vmClass = (Class<M>) actualTypeArguments[0];
        try {
            model = vmClass.newInstance();
            model.application = application;
        } catch (IllegalAccessException | InstantiationException e) {
            Log.i(TAG, e.toString());
        }
    }

    @Override
    public void onStart() {
        Log.i(TAG, "onStart");
    }

    @Override
    public void onResume() {
        Log.i(TAG, "onResume");
    }

    @Override
    public void onPause() {
        Log.i(TAG, "onPause");
    }

    @Override
    public void onStop() {
        Log.i(TAG, "onStop");
    }

    @Override
    public void onDestroy() {
        Log.i(TAG, "onDestroy");
    }
}

File: QuickUIMate/src/main/java/com/tomcan/frame/vm/QuickViewModel.kt
package com.tomcan.frame.vm

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import com.tomcan.frame.m.QuickModel
import com.tomcan.frame.obs.IBaseLifecycle
import java.lang.reflect.ParameterizedType

/**
 * @author TomCan
 * @description:
 * @date :2020/8/5 15:34
 */
abstract class QuickViewModel<M : QuickModel>(val context: Application) : AndroidViewModel(
    context
), IBaseLifecycle {
    val TAG: String = javaClass.simpleName
    lateinit var model: M

    override fun onAny(owner: LifecycleOwner?, event: Lifecycle.Event?) {
        Log.d(TAG, "onAny")
    }

    override fun onCreate() {
        Log.d(TAG, "onCreate")
        val genericSuperclass = checkNotNull(javaClass.genericSuperclass as ParameterizedType)
        val actualTypeArguments = genericSuperclass.actualTypeArguments
        val vmClass = actualTypeArguments[0] as Class<M>
        try {
            model = vmClass.newInstance()
            model.application = context
        } catch (e: IllegalAccessException) {
            Log.i(TAG, e.toString())
        } catch (e: InstantiationException) {
            Log.i(TAG, e.toString())
        }
    }

    override fun onStart() {
        Log.d(TAG, "onStart")
    }

    override fun onResume() {
        Log.d(TAG, "onResume")
    }

    override fun onPause() {
        Log.d(TAG, "onPause")
    }

    override fun onStop() {
        Log.d(TAG, "onStop")
    }

    override fun onDestroy() {
        Log.d(TAG, "onDestroy")
    }
}
